File: examples/cs/ex_curvesurfaceintersect.cs
using Rhino;
using Rhino.Geometry;
using Rhino.Geometry.Intersect;
using Rhino.Input.Custom;
using Rhino.DocObjects;
using Rhino.Commands;

namespace examples_cs
{
  [System.Runtime.InteropServices.Guid("D3E509D4-3791-42C1-A136-FFBA37359290")]
  public class CurveSurfaceIntersectCommand : Command
  {
    public override string EnglishName { get { return "csCurveSurfaceIntersect"; } }

    protected override Result RunCommand(RhinoDoc doc, RunMode mode)
    {
      var gs = new GetObject();
      gs.SetCommandPrompt("select brep");
      gs.GeometryFilter = ObjectType.Brep;
      gs.DisablePreSelect();
      gs.SubObjectSelect = false;
      gs.Get();
      if (gs.CommandResult() != Result.Success)
        return gs.CommandResult();
      var brep = gs.Object(0).Brep();

      var gc = new GetObject();
      gc.SetCommandPrompt("select curve");
      gc.GeometryFilter = ObjectType.Curve;
      gc.DisablePreSelect();
      gc.SubObjectSelect = false;
      gc.Get();
      if (gc.CommandResult() != Result.Success)
        return gc.CommandResult();
      var curve = gc.Object(0).Curve();

      if (brep == null || curve == null)
        return Result.Failure;

      var tolerance = doc.ModelAbsoluteTolerance;

      Point3d[] intersectionPoints;
      Curve[] overlapCurves;
      if (!Intersection.CurveBrep(curve, brep, tolerance, out overlapCurves, out intersectionPoints))
      {
        RhinoApp.WriteLine("curve brep intersection failed");
        return Result.Nothing;
      }

      foreach (var overlapCurve in overlapCurves)
        doc.Objects.AddCurve(overlapCurve);
      foreach (var intersectionPoint in intersectionPoints)
        doc.Objects.AddPoint(intersectionPoint);

      RhinoApp.WriteLine(string.Format("{0} overlap curves, and {1} intersection points", overlapCurves.Length, intersectionPoints.Length));
      doc.Views.Redraw();

      return Result.Success;
    }
  }
}

File: examples/py/ex_curvesurfaceintersect.py
import rhinoscriptsyntax as rs
from scriptcontext import *
import Rhino
import System.Collections.Generic as scg
import System as s

def RunCommand():
  srfid = rs.GetObject("select surface", rs.filter.surface | rs.filter.polysurface)
  if not srfid: return
 
  crvid = rs.GetObject("select curve", rs.filter.curve)
  if not crvid: return

  evs = rs.CurveSurfaceIntersection(crvid, srfid)

  if evs:
      addedObjs = []
      for ev in evs:
          if ev[0] == 2: #overlap
              crv = rs.coercecurve(crvid)
              if crv:
                  t0 = crv.ClosestPoint(ev[1])[1]
                  t1 = crv.ClosestPoint(ev[2])[1]
                  overlapCrv = crv.DuplicateCurve().Trim(t0,t1)
                  addedObjs.Add(doc.Objects.AddCurve(overlapCrv))
          else: #point
              addedObjs.Add(doc.Objects.AddPoint(ev[1]))

      if len(addedObjs) > 0:
          doc.Objects.Select.Overloads[scg.IEnumerable[s.Guid]](addedObjs)
  doc.Views.Redraw()

if __name__ == "__main__":
  RunCommand()
